### DISCUSSION TITLE
v2 FileDialog - UX

### DISCUSSION THREAD

**Discussion by tznind** [tui-cs/Terminal.Gui#3020] [Show and tell]
@tig rather than discuss in the PR https://github.com/gui-cs/Terminal.Gui/pull/2927#discussion_r1409313548 I thought I would go over some of the decisions I made when writing FileDialog.

I wanted to challenge myself to create the best ux possible for this dialog.

![image](https://github.com/gui-cs/Terminal.Gui/assets/31306100/61b88943-be51-481e-83e7-5e690f66b0fb)

I would like to ensure all this works correctly with the new key handling.  And if there is a way to work 'with the api' rather than against it that would be ideal.

# Caption in TextField

To save real estate the search box shows the 'label' within the box itself (as long as there is no text typed)

Windows
![image](https://github.com/gui-cs/Terminal.Gui/assets/31306100/21d8c02f-d7e1-48d4-9463-0a82aaf8f4d2)

Terminal
![image](https://github.com/gui-cs/Terminal.Gui/assets/31306100/e7152431-9a70-41ab-bddc-25b2d59673e3)

# Tab Completion

I wanted tab completion to work 'as good as on command line directly'.  This means when you start typing a path and hit <tab> it auto completes.

![image](https://github.com/gui-cs/Terminal.Gui/assets/31306100/23585bc9-1e6e-49f2-ac45-b2cd724bcf48)

If there are multiple things that match the Up/Down key - cycle those.

# Navigation

I wanted up/down/left/right to move you to a sensible part of the dialog.

So for example when in the PATH section pressing down should move to the Table.  Unless there are multiple autocomplete options to cycle in which case they should cycle.

There are other things like if you are at the top line of the table and press up the focus will move to the path.

I could go on but want to highlight the goal here which is to make the ux be reactive and sensible about how it interprets the key press based on its current state.

# Summary

Happy to talk through any and all of these decisions, identify any ways they could change to work better with new API for keypresses and/or they could inform the API development.

**Comment by tig** [comment]
To be clear, this comment...

> All of the navigation complexity in file dialog is complex and confusing. I'm curious why you didn't just depend on the tab stop tab index functionality.

was not abou the functionality or look & feel. I love it.'

It was about the code, esp the `NavigateTo` stuff and alll the event handlers.

I see all that and wonder "why doesn't the library just make that easy?"
